pyodide-build/pyodide_build/build_env.py

```diff
@@ -0,0 +1,230 @@
+import functools
+import os
+import subprocess
+import sys
+from contextlib import ExitStack, redirect_stdout
+from io import StringIO
+from pathlib import Path
+
+if sys.version_info < (3, 11, 0):
+    import tomli as tomllib
+else:
+    import tomllib
+
+from .logger import logger
+
+BUILD_VARS: set[str] = {
+    "PATH",
+    "PYTHONPATH",
+    "PYODIDE_ROOT",
+    "PYTHONINCLUDE",
+    "NUMPY_LIB",
+    "PYODIDE_PACKAGE_ABI",
+    "HOME",
+    "HOSTINSTALLDIR",
+    "TARGETINSTALLDIR",
+    "SYSCONFIG_NAME",
+    "HOSTSITEPACKAGES",
+    "PYVERSION",
+    "PYMAJOR",
+    "PYMINOR",
+    "PYMICRO",
+    "CPYTHONBUILD",
+    "CPYTHONLIB",
+    "SIDE_MODULE_CFLAGS",
+    "SIDE_MODULE_CXXFLAGS",
+    "SIDE_MODULE_LDFLAGS",
+    "STDLIB_MODULE_CFLAGS",
+    "WASM_LIBRARY_DIR",
+    "WASM_PKG_CONFIG_PATH",
+    "CARGO_BUILD_TARGET",
+    "CARGO_TARGET_WASM32_UNKNOWN_EMSCRIPTEN_LINKER",
+    "RUSTFLAGS",
+    "PYO3_CROSS_LIB_DIR",
+    "PYO3_CROSS_INCLUDE_DIR",
+    "PYODIDE_EMSCRIPTEN_VERSION",
+    "PLATFORM_TRIPLET",
+    "SYSCONFIGDATA_DIR",
+    "RUST_TOOLCHAIN",
+}
+
+
+def init_environment(*, quiet: bool = False) -> None:
+    """
+    Initialize Pyodide build environment.
+    This function needs to be called before any other Pyodide build functions.
+    """
+    if os.environ.get("__LOADED_PYODIDE_ENV"):
+        return
+
+    os.environ["__LOADED_PYODIDE_ENV"] = "1"
+
+    _set_pyodide_root(quiet=quiet)
+
+
+@functools.cache
+def get_pyodide_root() -> Path:
+    init_environment()
+    return Path(os.environ["PYODIDE_ROOT"])
+
+
+def in_xbuildenv() -> bool:
+    pyodide_root = get_pyodide_root()
+    return pyodide_root.name == "pyodide-root"
+
+
+def search_pyodide_root(curdir: str | Path, *, max_depth: int = 5) -> Path:
+    """
+    Recursively search for the root of the Pyodide repository,
+    by looking for the pyproject.toml file in the parent directories
+    which contains [tool.pyodide] section.
+    """
+
+    # We want to include "curdir" in parent_dirs, so add a garbage suffix
+    parent_dirs = (Path(curdir) / "garbage").parents[:max_depth]
+
+    for base in parent_dirs:
+        pyproject_file = base / "pyproject.toml"
+
+        if not pyproject_file.is_file():
+            continue
+
+        try:
+            with pyproject_file.open("rb") as f:
+                configs = tomllib.load(f)
+        except tomllib.TOMLDecodeError as e:
+            raise ValueError(f"Could not parse {pyproject_file}.") from e
+
+        if "tool" in configs and "pyodide" in configs["tool"]:
+            return base
+
+    raise FileNotFoundError(
+        "Could not find Pyodide root directory. If you are not in the Pyodide directory, set `PYODIDE_ROOT=<pyodide-root-directory>`."
+    )
+
+
+@functools.cache
+def get_build_environment_vars() -> dict[str, str]:
```

Comment:
Previously was: `common.set_build_environment`

Reply:
Could you please check that this includes changes from https://github.com/pyodide/pyodide/pull/3744 ?